#include "mainwindow.h"
#include "AddUserForm.h"
#include "ui_mainwindow.h"

//–í—ã–≤–æ–¥ –∏ –¥–∞–Ω–Ω—ã–µ
#include <QApplication>
#include <QDebug>
#include <QFile>
#include <QFileInfo>
#include <QSettings>
#include <QDate>
#include <QMessageBox>
#include <QTextCharFormat>

//–¥–ª—è —Ä–∞–±–æ—Ç—ã —Å –ø–æ—á—Ç–æ–π
#include <QDesktopServices>
#include <QUrl>
#include <QProcess>

// –¥–ª—è —Å—Ç–∏–ª–µ–π
#include <QStylePainter>
#include <QCalendarWidget>
#include <QHeaderView>
#include <QStyle>
#include <QPalette>
#include <QStyleFactory>
#include <QTableWidget>

#include <QContextMenuEvent>


MainWindow::MainWindow(QWidget *parent)
    : QMainWindow(parent)
    , ui(new Ui::MainWindow)
    , filePath("C:/temp/database.txt")
//    , filePath("O:/06. –î–∞–Ω–Ω—ã–µ –æ—Ç–¥–µ–ª–æ–≤/–ù–ò–û –ú–ú–ù–ú/–†–∞–±–æ—á–∏–µ –¥–∞–Ω–Ω—ã–µ/–°–æ–±—Å—Ç–≤–µ–Ω–Ω—ã–µ –ø—Ä–æ–≥—Ä–∞–º–º—ã/Murzin_AV/Base/birthdays.txt")
{
    ui->setupUi(this);
    init();
    readSettings();

    ui->tableView_Vacation->setHidden(!isVacation);
    ui->cb_Vacation->setHidden(!isVacation);
    ui->calendarWidget->setHidden(isVacation || isChat);
    ui->tb_Vacation->setChecked(isVacation);
    ui->tb_Chat->setChecked(isChat);
    ui->chatWidget->setHidden(!isChat);

    on_tb_Style_clicked(isStyle);
    ui->tb_Style->setChecked(isStyle);

}

MainWindow::~MainWindow()
{
    delete ui;
}

void MainWindow::dateOgonki()
{
    QDate currentDate = QDate::currentDate();

    for ( int i = 0; i < eModel->rowCount(); i++ ) {
        QModelIndex index = eModel->index(i, EmploeeModel::EColumn::Birthday);
        QString birthday = eModel->data(index, Qt::EditRole).toString();

        QStringList date = birthday.split('.');
        QDate eventDate(currentDate.year(), date[1].toInt(), date[0].toInt());

        index = eModel->index(i, EmploeeModel::EColumn::Name);
        QString name = eModel->data(index, Qt::EditRole).toString();
        index = eModel->index(i, EmploeeModel::EColumn::Famili);
        QString famili = eModel->data(index, Qt::EditRole).toString();
        index = eModel->index(i, EmploeeModel::EColumn::Otch);
        QString otch = eModel->data(index, Qt::EditRole).toString();

        dateLight(eventDate, currentDate, QString(famili+" "+ name+" "+ otch));
    }
}

void MainWindow::dateLight(const QDate &eventDate, const QDate &currentDate, const QString &toolText, QColor color, bool isAppend)
{
    Q_UNUSED(currentDate);
    QTextCharFormat textCharFormat = ui->calendarWidget->dateTextFormat(eventDate);
    textCharFormat.setToolTip( isAppend ? textCharFormat.toolTip() + toolText : toolText + " üéÇ");
    textCharFormat.setForeground(color);

    ui->calendarWidget->setDateTextFormat(eventDate, textCharFormat);
}

void MainWindow::removeLineFromFile(const QString &filename, int lineIndex)
{
    QFile file(filename);
    if ( file.open(QIODevice::ReadWrite) ) {
        QTextStream stream(&file);
        QString line;
        QStringList lines;

        // –ß–∏—Ç–∞–µ–º –≤—Å–µ —Å—Ç—Ä–æ–∫–∏ –∏–∑ —Ñ–∞–π–ª–∞
        while ( !stream.atEnd() ) {
            line = stream.readLine();
            lines.append(line);
        }

        // –£–¥–∞–ª—è–µ–º —Å—Ç—Ä–æ–∫—É —Å —É–∫–∞–∑–∞–Ω–Ω—ã–º –∏–Ω–¥–µ–∫—Å–æ–º
        if ( lineIndex >= 0 && lineIndex < lines.size() ) {
            lines.removeAt(lineIndex);
        }

        // –ü–µ—Ä–µ–∑–∞–ø–∏—Å—ã–≤–∞–µ–º —Ñ–∞–π–ª —Å –æ–±–Ω–æ–≤–ª–µ–Ω–Ω—ã–º–∏ —Å—Ç—Ä–æ–∫–∞–º–∏
        file.resize(0);
        for ( const QString &updatedLine : lines ) {
            stream << updatedLine << Qt::endl;
        }
    }
    file.close();
}

void MainWindow::handleHeaderClicked(int logicalIndex)
{

    Qt::SortOrder newSortOrder = (eModel->sortOrder == Qt::AscendingOrder) ? Qt::DescendingOrder : Qt::AscendingOrder;
    eModel->sort(logicalIndex, newSortOrder);
    eModel->layoutChanged();

}

bool MainWindow::confirmDeletion()
{
    QMessageBox::StandardButton reply;
    reply = QMessageBox::question(this, "–ü–æ–¥—Ç–≤–µ—Ä–∂–¥–µ–Ω–∏–µ", "–í—ã —É–≤–µ—Ä–µ–Ω—ã, —á—Ç–æ —Ö–æ—Ç–∏—Ç–µ —É–¥–∞–ª–∏—Ç—å —ç—Ç—É –∑–∞–ø–∏—Å—å?",
                                  QMessageBox::Yes | QMessageBox::No);
    return reply == QMessageBox::Yes;
}

void MainWindow::removeRowFromModel(int row)
{
    eModel->removeRow(row);
    eModel->layoutChanged();
    ui->tableView->setCurrentIndex(QModelIndex());
    ui->tableView->update();
}

void MainWindow::removeRowFromFile(int row)
{
    QString fileName = filePath;
    QFile file(fileName);
    if ( file.open(QIODevice::ReadWrite | QIODevice::Text) ) {
        QTextStream in(&file);
        QStringList lines;
        while ( !in.atEnd() ) {
            QString line = in.readLine();
            lines.append(line);
        }
        file.resize(0); // –û—á–∏—â–∞–µ–º —Ñ–∞–π–ª

        for ( const QString &line : lines ) {
            QStringList words = line.split('\t');
            if ( words.isEmpty() ) continue;

            int currentRow = words[0].toInt();
            if ( currentRow != row ) {
                in << line << Qt::endl;
            }
        }
        file.close();
    }
}

void MainWindow::contextMenuEvent(QContextMenuEvent *event)
{
    QSettings settings("C:\\Temp\\config.cfg", QSettings::IniFormat);
    QMenu menu(this);
    for ( int col = 0; col < eModel->columnCount(); col++ ) {
        QAction *action = new QAction(eModel->headerData(col, Qt::Horizontal).toString(), this);
        action->setCheckable(true);
        if ( settings.value(eModel->headerData(col, Qt::Horizontal).toString()).toString().isEmpty() ) {
            action->setChecked(true);
            ui->tableView->setColumnHidden(col, false);
        }
        else {
            action->setChecked(settings.value(eModel->headerData(col, Qt::Horizontal).toString()).toBool());
            ui->tableView->setColumnHidden(col, !settings.value(eModel->headerData(col, Qt::Horizontal).toString()).toBool());
        }
        menu.addAction(action);

        connect(action, &QAction::toggled, this, [this, col, &settings] (bool checked) {
            settings.setValue(eModel->headerData(col, Qt::Horizontal).toString(), checked);
            ui->tableView->setColumnHidden(col, !checked);
        });
    }
    menu.exec(event->globalPos());
}

void MainWindow::showEvent(QShowEvent *event)
{
    QMainWindow::showEvent(event);

    // –í–æ—Å—Å—Ç–∞–Ω–æ–≤–∏—Ç–µ —Å–æ—Å—Ç–æ—è–Ω–∏–µ —Å–ø–ª–∏—Ç—Ç–µ—Ä–∞ –ø—Ä–∏ –æ—Ç–∫—Ä—ã—Ç–∏–∏ –ø—Ä–∏–ª–æ–∂–µ–Ω–∏—è
    QSettings settings("C:\\Temp\\config.cfg", QSettings::IniFormat);
    QByteArray splitterState = settings.value("splitter/geometry").toByteArray();
    if (!splitterState.isEmpty()) {
        ui->splitter->restoreState(splitterState);
    }
    splitterState = settings.value("splitter/ChatGeometry").toByteArray();
    if (!splitterState.isEmpty()) {
        ui->chatWidget->setSplitter(splitterState);
    }
    // –í–æ—Å—Å—Ç–∞–Ω–æ–≤–ª–µ–Ω–∏–µ —Ä–∞–∑–º–µ—Ä–æ–≤ –æ–∫–Ω–∞
    QByteArray windowGeometry = settings.value("window/geometry").toByteArray();
    if (!windowGeometry.isEmpty()) {
        restoreGeometry(windowGeometry);
    }
}

void MainWindow::closeEvent(QCloseEvent *event)
{
    QSettings settings("C:\\Temp\\config.cfg", QSettings::IniFormat);
    // –°–æ—Ö—Ä–∞–Ω–µ–Ω–∏–µ —Å–æ—Å—Ç–æ—è–Ω–∏—è —Å–ø–ª–∏—Ç—Ç–µ—Ä–∞
    settings.setValue("splitter/geometry", ui->splitter->saveState());
    // –°–æ—Ö—Ä–∞–Ω–µ–Ω–∏–µ —Ä–∞–∑–º–µ—Ä–æ–≤ –æ–∫–Ω–∞
    settings.setValue("window/geometry", saveGeometry());
    // –°–æ—Ö—Ä–∞–Ω–µ–Ω–∏–µ —Å–æ—Å—Ç–æ—è–Ω–∏—è —Å–ø–ª–∏—Ç—Ç–µ—Ä–∞ —É ChatWidget
    settings.setValue("splitter/ChatGeometry", ui->chatWidget->getSplitter()->saveState());

    QMainWindow::closeEvent(event);
}

void MainWindow::init()
{
    eModel = new EmploeeModel(this);
    ui->tableView->setModel(eModel);

    vModel = new VacationModel(eModel, ui->cb_Vacation, this);
    vacation(ui->cb_Vacation->currentIndex() + 1);
    QObject::connect(ui->cb_Vacation, static_cast<void(QComboBox::*)(int)>(&QComboBox::currentIndexChanged), [&] (int index) {
        vacation(index + 1);
    });
    ui->tableView_Vacation->setModel(vModel);

    connect(ui->tableView->horizontalHeader(), &QHeaderView::sectionClicked, this, &MainWindow::handleHeaderClicked);

    dateOgonki();
    connect(ui->tableView->model(), &QAbstractTableModel::dataChanged, [&]() {
        dateOgonki();
        if ( senderSignalIndex() == EmploeeModel::EColumn::Birthday )  {
//TODO –ø—Ä–∏ –ª—é–±–æ–º –∏–∑–º–µ–Ω–µ–Ω–∏–µ –≤—ã–∑—ã–≤–∞–µ—Ç—Å—è dataOgonki (–Ω–µ –ø—Ä–∞–ª—å–Ω–æ!!!!!)
        }

    });

    ui->cb_Vacation->setCurrentIndex(QDate::currentDate().month()-1);
    ui->tableView_Vacation->horizontalHeader()->setSectionResizeMode(QHeaderView::ResizeToContents);

    ui->tableView->horizontalHeader()->setSectionResizeMode(QHeaderView::ResizeToContents);
    ui->tableView->horizontalHeader()->setStretchLastSection(true);

    GenderDelegate *delegate = new GenderDelegate(this);
    ui->tableView->setItemDelegateForColumn(EmploeeModel::Gender, delegate);

//    VacationDelegate *vDelegate = new VacationDelegate(vModel->getEmploeeVacationList(), this);
//    ui->tableView_Vacation->setItemDelegate(vDelegate);
}

void MainWindow::readSettings()
{
    QSettings settings("C:\\Temp\\config.cfg", QSettings::IniFormat);
    for ( int col = 0; col < eModel->columnCount(); col++ ) {
        if ( !settings.value(eModel->headerData(col, Qt::Horizontal).toString()).toString().isEmpty() ) {
            ui->tableView->setColumnHidden(col, !settings.value(eModel->headerData(col, Qt::Horizontal).toString()).toBool());
        }
    }

    isStyle    = settings.value("style").toBool();
    isVacation = settings.value("isVacation").toBool();
    isChat     = settings.value("isChat").toBool();
}

void MainWindow::vacation(int month)
{
    vModel->clear();
    // –û—Ç–∫—Ä—ã–≤–∞–µ–º —Ñ–∞–π–ª
//    QFile file("O:/06. –î–∞–Ω–Ω—ã–µ –æ—Ç–¥–µ–ª–æ–≤/–ù–ò–û –ú–ú–ù–ú/–†–∞–±–æ—á–∏–µ –¥–∞–Ω–Ω—ã–µ/–°–æ–±—Å—Ç–≤–µ–Ω–Ω—ã–µ –ø—Ä–æ–≥—Ä–∞–º–º—ã/Murzin_AV/Base/Vacation.txt");
    QFile file("C:/temp/Vacation.txt");
    if ( ! file.open(QIODevice::ReadOnly | QIODevice::Text) ) return;

    QStringList familiList;
    for ( int i = 0; i < eModel->rowCount(); i++ ) {
        familiList.push_back(eModel->data(QModelIndex(eModel->index(i, EmploeeModel::Famili))).toString() + " " +
                             eModel->data(QModelIndex(eModel->index(i, EmploeeModel::Name))).toString() + " " +
                             eModel->data(QModelIndex(eModel->index(i, EmploeeModel::Otch))).toString()
                             );
    }


    // –ß–∏—Ç–∞–µ–º –∏–Ω—Ñ–æ—Ä–º–∞—Ü–∏—é –∏–∑ —Ñ–∞–π–ª–∞
    QTextStream in(&file);
    while ( !in.atEnd() ) {
        QStringList splitLine = in.readLine().split('\t');

        //TODO –ø–µ—Ä–≤–æ–µ —á–∏—Å–ª–æ –Ω–µ –∫–æ—Ä–µ–∫—Ç–Ω–æ !!!
        int it = 0;
        for ( const QString &word : splitLine ) {
            if ( familiList.contains(word) ) {
                EmploeeVacation *emploeeVacation = new EmploeeVacation;
                emploeeVacation->fio = word;

                QDate firstDay = QDate::fromString(splitLine[it+1], "dd.MM.yyyy");
                QDate lastDay  = QDate::fromString(splitLine[it+2], "dd.MM.yyyy");

                if ( firstDay.isValid() && firstDay.month() == month ) {
                    QString days = splitLine[it+3];
                    emploeeVacation->type = splitLine[it+4];

                    int haveDays = days.toInt();
                    for (int i = 0; i <= firstDay.daysInMonth() -1; i++ ) {
                        if ( firstDay.day() > i )
                            emploeeVacation->dayList.push_back(false);
                        else {
                            if ( haveDays >= 0 ) emploeeVacation->dayList.push_back(true);
                            else                 emploeeVacation->dayList.push_back(false);

                            haveDays--;
                        }
                    }
                    vModel->addItem(emploeeVacation);
                }

                if ( (firstDay.isValid() && firstDay.month() != month) && (lastDay.isValid() && lastDay.month() == month) ) {
                    for ( int i = 0; i <= lastDay.daysInMonth() -1; i++ ) {
                        if ( lastDay.day() > i ) emploeeVacation->dayList.push_back(true);
                        else                     emploeeVacation->dayList.push_back(false);
                    }
                    vModel->addItem(emploeeVacation);
                }

                //TODO –ø–µ—Ä–µ–Ω–µ—Å—Ç–∏ –ø–æ–∫—Ä–∞—Å–∫—É (—Ä–µ–∞–ª–∏–∑–æ–≤–∞—Ç—å –≤—ã–±–æ—Ä —Ü–≤–µ—Ç–∞ –ø–æ–ª—å–∑–æ–≤–∞—Ç–µ–ª—å—Å–∫–∏  –∏–∑ –ø—Ä–∏–ª–æ–∂–µ–Ω–∏—è)
                if (word.startsWith("–ú—É—Ä–∑–∏–Ω") ) {
                    emploeeVacation->color = *new QColor(255, 0, 204);
                }
                else if (word.startsWith("–Ø–∫–æ–≤–ª–µ–≤" ) ) {
                    emploeeVacation->color = *new QColor(61, 208, 154);
                }
                else if (word.startsWith("–ú–∞–ª—é–≥–∏–Ω" ) ) {
                    emploeeVacation->color = *new QColor(0,0,255);
                }
                else if (word.startsWith("–ñ—É–ª–∏–Ω" ) ) {
                    emploeeVacation->color = *new QColor(30,33,61);
                }
                else if (word.startsWith("–ö–∞–∑–Ω–∞—á–µ–µ–≤" ) ) {
                    emploeeVacation->color = *new QColor(169,29,17);
                }
                else if (word.startsWith("–ê–Ω–¥—Ä–µ–µ–≤" ) ) {
                    emploeeVacation->color = *new QColor(169,131,7);
                }
                else if (word.startsWith("–ú–∞–ª—ã—à–µ–≤–∞" ) ) {
                    emploeeVacation->color = *new QColor(147,170,0);
                }
                else if (word.startsWith("–ë–µ–ª–æ—É—Å–æ–≤–∞" ) ) {
                    emploeeVacation->color = *new QColor(76,187,23);
                }
                else if (word.startsWith("–ö–æ–Ω–¥—Ä–∞—Ç—å–µ–≤" ) ) {
                    emploeeVacation->color = *new QColor(0,255,255);
                }
                else if (word.startsWith("–†—ã–±–∞—á—É–∫" ) ) {
                    emploeeVacation->color = *new QColor(150,186,181);
                }
                else if (word.startsWith("–ö–æ–Ω—è—Ö–∏–Ω–∞" ) ) {
                    emploeeVacation->color = *new QColor(102,0,153);
                }
                else if (word.startsWith("–†–µ–∑–∞–Ω–æ–≤" ) ) {
                    emploeeVacation->color = *new QColor(19, 10, 143);
                }
                else if (word.startsWith("–ö–∞—Ä–ø–æ–≤" ) ) {
                    emploeeVacation->color = *new QColor(19, 136, 8);
                }
            }
            it++;
        }
    }

    vModel->layoutChanged();
    VacationDelegate *vDelegate = new VacationDelegate(vModel->getEmploeeVacationList(), this);
    ui->tableView_Vacation->setItemDelegate(vDelegate);
}

void MainWindow::on_pb_Add_clicked()
{
    // –°–æ–∑–¥–∞–Ω–∏–µ –∏ –æ—Ç–∫—Ä—ã—Ç–∏–µ –Ω–æ–≤–æ–π —Ñ–æ—Ä–º—ã –¥–ª—è –¥–æ–±–∞–≤–ª–µ–Ω–∏—è –¥–∞–Ω–Ω—ã—Ö
    Emploee *emploee = new Emploee;
    AddUserForm *addUserForm = new AddUserForm(*emploee);
    addUserForm->setWindowFlags(addUserForm->windowFlags() & ~Qt::WindowContextHelpButtonHint);

    if ( addUserForm->exec() == QDialog::Accepted ) {
        // –î–æ–±–∞–≤–ª–µ–Ω–∏–µ –¥–∞–Ω–Ω—ã—Ö –≤ –º–æ–¥–µ–ª—å
        eModel->addData(emploee);
        dateOgonki();
    }
}

void MainWindow::on_pb_Del_clicked()
{
    int row = ui->tableView->currentIndex().row();
    if ( row >= 0 ) {
        if ( confirmDeletion() ) {
            removeRowFromModel(row);
            removeRowFromFile(row);
        }
    }
}

void MainWindow::on_pb_Save_clicked()
{
    QFile file(filePath);
    if ( ! file.open(QIODevice::WriteOnly | QIODevice::Text) )return;

    QTextStream out(&file);

    for ( int row = 0; row < eModel->rowCount(); row++ ) {
        for ( int col = 0; col < eModel->columnCount(); col++ ) {
            QModelIndex index = eModel->index(row, col);
            out << eModel->data(index).toString() + '\t';
        }
        out << Qt::endl;
    }
    file.close();
}

void MainWindow::on_pb_Sort_clicked()
{
    if ( sortClickCount % 2 == 0) {
        eModel->sort(EmploeeModel::Birthday, Qt::AscendingOrder, true);
        ui->pb_Sort->setIcon(QIcon(":/src/resources/Icon/sort-ascending.png"));
    }
    else {
        eModel->sort(EmploeeModel::Birthday, Qt::DescendingOrder, true);
        ui->pb_Sort->setIcon(QIcon(":/src/resources/Icon/sort-descending.png"));
    }
    sortClickCount++;
    eModel->layoutChanged();
}

void MainWindow::on_pb_Send_clicked()
{
    QModelIndexList selectedIndexes = ui->tableView->selectionModel()->selectedIndexes();
    QStringList emailAddresses;

    // –ü–æ–ª—É—á–∞–µ–º –ø–æ—á—Ç–æ–≤—ã–µ –∞–¥—Ä–µ—Å–∞ –∏–∑ –≤—ã–¥–µ–ª–µ–Ω–Ω—ã—Ö —è—á–µ–µ–∫
    for (const QModelIndex& index : selectedIndexes)
    {
        if (index.column() == EmploeeModel::Email)
        {
            QString emailAddress = eModel->data(index, Qt::DisplayRole).toString();
            emailAddresses.append(emailAddress);
        }
    }

    // –ü—Ä–æ–≤–µ—Ä—è–µ–º, –≤—ã–±—Ä–∞–Ω–∞ –ª–∏ —Ö–æ—Ç—è –±—ã –æ–¥–Ω–∞ —ç–ª–µ–∫—Ç—Ä–æ–Ω–Ω–∞—è –ø–æ—á—Ç–∞
    if (emailAddresses.isEmpty())
    {
        QMessageBox::warning(this, "–ü—Ä–µ–¥—É–ø—Ä–µ–∂–¥–µ–Ω–∏–µ", "–ù–µ –≤—ã–±—Ä–∞–Ω–∞ —ç–ª–µ–∫—Ç—Ä–æ–Ω–Ω–∞—è –ø–æ—á—Ç–∞!");
        return;
    }

    // –§–æ—Ä–º–∏—Ä—É–µ–º —Å—Ç—Ä–æ–∫—É —Å –ø–æ—á—Ç–æ–≤—ã–º–∏ –∞–¥—Ä–µ—Å–∞–º–∏
    QString emailString = emailAddresses.join(";");

    // –ü—Ä–æ–≤–µ—Ä—è–µ–º, —è–≤–ª—è–µ—Ç—Å—è –ª–∏ —Å—Ç–∞–Ω–¥–∞—Ä—Ç–Ω—ã–π –ø–æ—á—Ç–æ–≤—ã–π –∫–ª–∏–µ–Ω—Ç Outlook
    QSettings settings("HKEY_CURRENT_USER\\Software\\Clients\\Mail", QSettings::NativeFormat);
    QString defaultClient = settings.value("Default").toString();
    if ( defaultClient != "Microsoft Outlook" ) {
        QMessageBox::StandardButton reply = QMessageBox::question(this,
                                                                  "–ü—Ä–µ–¥—É–ø—Ä–µ–∂–¥–µ–Ω–∏–µ",
                                                                  "–î–ª—è –ø—Ä–æ–¥–æ–ª–∂–µ–Ω–∏—è —Ä–∞–±–æ—Ç—ã –Ω—É–∂–Ω–æ –∏—Å–ø–æ–ª—å–∑–æ–≤–∞—Ç—å Microsoft Outlook. "
                                                                  "–•–æ—Ç–∏—Ç–µ —Å–¥–µ–ª–∞—Ç—å –µ–≥–æ —Å—Ç–∞–Ω–¥–∞—Ä—Ç–Ω—ã–º –ø–æ—á—Ç–æ–≤—ã–º –∫–ª–∏–µ–Ω—Ç–æ–º?", QMessageBox::Yes|QMessageBox::No);
        if ( reply == QMessageBox::Yes ) {
            // –£—Å—Ç–∞–Ω–∞–≤–ª–∏–≤–∞–µ–º Outlook –≤ –∫–∞—á–µ—Å—Ç–≤–µ —Å—Ç–∞–Ω–¥–∞—Ä—Ç–Ω–æ–≥–æ –ø–æ—á—Ç–æ–≤–æ–≥–æ –∫–ª–∏–µ–Ω—Ç–∞
            settings.setValue("Default", "Microsoft Outlook");
        } else return; // –ü—Ä–µ–∫—Ä–∞—â–∞–µ–º –æ—Ç–ø—Ä–∞–≤–∫—É –ø–æ—á—Ç—ã
    }

    // –ü–æ–ª—É—á–∞–µ–º –¥–∞–Ω–Ω—ã–µ –æ –±—É–¥—É—â–µ–º –∏–º–µ–Ω–∏–Ω–Ω–∏–∫–µ
    QModelIndex index = ui->tableView->currentIndex();
    QString name = eModel->data(index.sibling(index.row(), EmploeeModel::Name), Qt::DisplayRole).toString();
    QString famili = eModel->data(index.sibling(index.row(), EmploeeModel::Famili), Qt::DisplayRole).toString();
    QString otch = eModel->data(index.sibling(index.row(), EmploeeModel::Otch), Qt::DisplayRole).toString();
    QString birthday = eModel->data(index.sibling(index.row(), EmploeeModel::Birthday), Qt::DisplayRole).toString();

    // –§–æ—Ä–º–∏—Ä—É–µ–º —Ç–µ–º—É –ø–∏—Å—å–º–∞
    QString subject = " &Subject=–î–µ–Ω—å —Ä–æ–∂–¥–µ–Ω–∏—è —É ¬´" + famili + " " + name + " " + otch + "¬ª";

    // –§–æ—Ä–º–∏—Ä—É–µ–º —Ç–µ–∫—Å—Ç –ø–∏—Å—å–º–∞
    QDate currentDate = QDate::currentDate();
    QStringList dateParts = birthday.split('.');
    //    QDate eventDate2 = QDate::fromString(birthday, "dd.MM.yyyy");
    QDate eventDate(currentDate.year(), dateParts[1].toInt(), dateParts[0].toInt());
    int age = currentDate.year() - QDate::fromString(birthday, "dd.MM.yyyy").year();

    QString body = " &Body=–ß–µ—Ä–µ–∑ " + QString::number(currentDate.daysTo(eventDate)) +
            " –¥–Ω–µ–π —É ¬´" + famili + " " + name + " " + otch +
            "¬ª –±—É–¥–µ—Ç –¥–µ–Ω—å —Ä–æ–∂–¥–µ–Ω–∏—è, –µ–º—É –∏—Å–ø–æ–ª–Ω–∏—Ç—Å—è " + QString::number(age) + " üéÇ.";

    //    // –û—Ç–∫—Ä—ã–≤–∞–µ–º –ø–æ—á—Ç–æ–≤—ã–π –∫–ª–∏–µ–Ω—Ç —Å –ø—Ä–µ–¥–≤–∞—Ä–∏—Ç–µ–ª—å–Ω–æ –∑–∞–ø–æ–ª–Ω–µ–Ω–Ω—ã–º–∏ –∞–¥—Ä–µ—Å–∞–º–∏ —ç–ª–µ–∫—Ç—Ä–æ–Ω–Ω–æ–π –ø–æ—á—Ç—ã
    QDesktopServices::openUrl(QUrl("mailto:" + emailString+ subject +body));
}



void MainWindow::on_tb_Style_clicked(bool checked)
{
    QSettings settings("C:\\Temp\\config.cfg", QSettings::IniFormat);
    settings.setValue("style", checked);

    QApplication::setStyle("Fusion");
    ui->calendarWidget->setVerticalHeaderFormat(QCalendarWidget::NoVerticalHeader);

    QString style = " ";
    QString styleCalend = " ";
    QPalette palette = QApplication::style()->standardPalette();

    if ( !checked ) {
        // –ù–∞—Å—Ç—Ä–æ–π–∫–∞ —Ç–µ–º—ã –¥–ª—è —Ç—ë–º–Ω–æ–≥–æ —Å—Ç–∏–ª—è
        palette.setColor(QPalette::Window, QColor(53, 53, 53));
        palette.setColor(QPalette::WindowText, Qt::white);
        palette.setColor(QPalette::Base, QColor(25, 25, 25));
        palette.setColor(QPalette::AlternateBase, QColor(53, 53, 53));
        palette.setColor(QPalette::ToolTipBase, Qt::white);
        palette.setColor(QPalette::ToolTipText, Qt::white);
        palette.setColor(QPalette::Text, Qt::white);
        palette.setColor(QPalette::Button, QColor(53, 53, 53));
        palette.setColor(QPalette::ButtonText, Qt::white);
        palette.setColor(QPalette::BrightText, Qt::red);
        palette.setColor(QPalette::Link, QColor(42, 130, 218));
        //        QApplication::setPalette(palette);

        QFile styleFile(":/src/resources/Styles/calendar_style.qss");
        if ( styleFile.open(QIODevice::ReadOnly | QIODevice::Text) ) {
            styleCalend = styleFile.readAll();
            styleFile.close();
            //            ui->calendarWidget->setStyleSheet(style);
        }

        style = "QTableView:!active {"
                "  selection-background-color: #191919;"
                "  selection-color: white;"
                "}"
                "QTableView:active {"
                "  selection-background-color: #353535;"
                "  selection-color: white;"
                "}";

        //        ui->tableView->setStyleSheet(style);
        ui->tb_Style->setIcon(QIcon(":/src/resources/Icon/sun.png"));
    }
    else {
        ui->tb_Style->setIcon(QIcon(":/src/resources/Icon/sunBlack.png"));
    }

    QApplication::setPalette(palette);
    ui->calendarWidget->setStyleSheet(styleCalend);
    ui->tableView->setStyleSheet(style);
    style = "text-align:left; padding-left: 3px; padding-right: 3px";
    ui->pb_Add->setStyleSheet(style);
    ui->pb_Del->setStyleSheet(style);
    ui->pb_Save->setStyleSheet(style);
    ui->pb_Sort->setStyleSheet(style);
    ui->pb_Send->setStyleSheet(style);

}

void MainWindow::on_tb_Vacation_clicked(bool checked)
{
    ui->calendarWidget->setHidden(checked);
    ui->tableView_Vacation->setHidden(!checked);
    ui->cb_Vacation->setHidden(!checked);
    ui->chatWidget->setHidden(true);
    ui->tb_Chat->setChecked(false);

    QSettings settings("C:\\Temp\\config.cfg", QSettings::IniFormat);
    settings.setValue("isVacation", checked);
    settings.setValue("isChat", false);

}


void MainWindow::on_tb_Chat_clicked(bool checked)
{
    ui->calendarWidget->setHidden(checked);
    ui->tableView_Vacation->setHidden(true);
    ui->cb_Vacation->setHidden(true);
    ui->chatWidget->setHidden(!checked);
    ui->tb_Vacation->setChecked(false);

    isVacation = false;

    QSettings settings("C:\\Temp\\config.cfg", QSettings::IniFormat);
    settings.setValue("isChat", checked);
    settings.setValue("isVacation", false);
}


















//TODO
/*
void MainWindow::on_pb_Send_Shadow_clicked()
{
    QModelIndexList selectedIndexes = ui->tableView->selectionModel()->selectedIndexes();
    QStringList emailAddresses;

    // –ü–æ–ª—É—á–∞–µ–º –ø–æ—á—Ç–æ–≤—ã–µ –∞–¥—Ä–µ—Å–∞ –∏–∑ –≤—ã–¥–µ–ª–µ–Ω–Ω—ã—Ö —è—á–µ–µ–∫
    for (const QModelIndex& index : selectedIndexes)
    {
        if (index.column() == EmploeeModel::Email)
        {
            QString emailAddress = eModel->data(index, Qt::DisplayRole).toString();
            emailAddresses.append(emailAddress);
        }
    }

    // –ü—Ä–æ–≤–µ—Ä—è–µ–º, –≤—ã–±—Ä–∞–Ω–∞ –ª–∏ —Ö–æ—Ç—è –±—ã –æ–¥–Ω–∞ —ç–ª–µ–∫—Ç—Ä–æ–Ω–Ω–∞—è –ø–æ—á—Ç–∞
    if (emailAddresses.isEmpty())
    {
        QMessageBox::warning(this, "–ü—Ä–µ–¥—É–ø—Ä–µ–∂–¥–µ–Ω–∏–µ", "–ù–µ –≤—ã–±—Ä–∞–Ω–∞ —ç–ª–µ–∫—Ç—Ä–æ–Ω–Ω–∞—è –ø–æ—á—Ç–∞!");
        return;
    }

    // –§–æ—Ä–º–∏—Ä—É–µ–º —Å—Ç—Ä–æ–∫—É —Å –ø–æ—á—Ç–æ–≤—ã–º–∏ –∞–¥—Ä–µ—Å–∞–º–∏
    QString emailString = emailAddresses.join(";");

    // –§–æ—Ä–º–∏—Ä—É–µ–º —Ç–µ–∫—Å—Ç –ø–∏—Å—å–º–∞
    QString body = "–¢–µ–∫—Å—Ç –í–∞—à–µ–≥–æ —Å–æ–æ–±—â–µ–Ω–∏—è";
    QString subject = "–¢–µ–º–∞ –í–∞—à–µ–≥–æ —Å–æ–æ–±—â–µ–Ω–∏—è";

    // –°–æ–∑–¥–∞–µ–º –∑–∞–¥–∞—á—É –¥–ª—è –æ—Ç–ø—Ä–∞–≤–∫–∏ –ø–æ—á—Ç—ã
    SendEmailTask* task = new SendEmailTask(emailString, subject, body);

    // –û—Ç–ø—Ä–∞–≤–ª—è–µ–º –∑–∞–¥–∞—á—É –≤ –ø—É–ª –ø–æ—Ç–æ–∫–æ–≤
    QThreadPool::globalInstance()->start(task);
}

//               —Ä–∞–±–æ—á–∏–π –º–µ—Ç–æ–¥!!!!
//void MainWindow::on_pb_Send_Shadow_clicked()
//{
//    QModelIndexList selectedIndexes = ui->tableView->selectionModel()->selectedIndexes();
//    QStringList emailAddresses;

//    // –ü–æ–ª—É—á–∞–µ–º –ø–æ—á—Ç–æ–≤—ã–µ –∞–¥—Ä–µ—Å–∞ –∏–∑ –≤—ã–¥–µ–ª–µ–Ω–Ω—ã—Ö —è—á–µ–µ–∫
//    for (const QModelIndex& index : selectedIndexes)
//    {
//        if (index.column() == EmploeeModel::Email)
//        {
//            QString emailAddress = eModel->data(index, Qt::DisplayRole).toString();
//            emailAddresses.append(emailAddress);
//        }
//    }

//    // –ü—Ä–æ–≤–µ—Ä—è–µ–º, –≤—ã–±—Ä–∞–Ω–∞ –ª–∏ —Ö–æ—Ç—è –±—ã –æ–¥–Ω–∞ —ç–ª–µ–∫—Ç—Ä–æ–Ω–Ω–∞—è –ø–æ—á—Ç–∞
//    if (emailAddresses.isEmpty())
//    {
//        QMessageBox::warning(this, "–ü—Ä–µ–¥—É–ø—Ä–µ–∂–¥–µ–Ω–∏–µ", "–ù–µ –≤—ã–±—Ä–∞–Ω–∞ —ç–ª–µ–∫—Ç—Ä–æ–Ω–Ω–∞—è –ø–æ—á—Ç–∞!");
//        return;
//    }

//    // –§–æ—Ä–º–∏—Ä—É–µ–º —Å—Ç—Ä–æ–∫—É —Å –ø–æ—á—Ç–æ–≤—ã–º–∏ –∞–¥—Ä–µ—Å–∞–º–∏
//    QString emailString = emailAddresses.join(";");

//    // –§–æ—Ä–º–∏—Ä—É–µ–º —Ç–µ–∫—Å—Ç –ø–∏—Å—å–º–∞
//    QString body = "–¢–µ–∫—Å—Ç –≤–∞—à–µ–≥–æ —Å–æ–æ–±—â–µ–Ω–∏—è";
//    QString subject = "–¢–µ–º–∞ –≤–∞—à–µ–≥–æ —Å–æ–æ–±—â–µ–Ω–∏—è";

//    // –û—Ç–∫—Ä—ã—Ç—å –ø–æ—á—Ç–æ–≤—ã–π –∫–ª–∏–µ–Ω—Ç –ø–æ —É–º–æ–ª—á–∞–Ω–∏—é –∏ –∑–∞–ø–æ–ª–Ω–∏—Ç—å –ø–æ–ª—è
//    QDesktopServices::openUrl(QUrl("mailto:" + emailString + "?subject=" + subject + "&body=" + body));
//}

*/



